### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/Matrix/ToLin): for R-module M, End(Mⁿ) ≅ Matₙₓₙ(End(M))

### PR DESCRIPTION
**Pull request by jjaassoonn**
This isomorphism is defined in a similar fashion to [LinearMap.toMatrix'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Matrix/ToLin.html#LinearMap.toMatrix'), but we consider End(Mⁿ) instead of End(Rⁿ)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Matrix/Basic.lean [hunk 1]
```diff
@@ -1497,6 +1497,19 @@ theorem mapMatrix_trans (f : α ≃+* β) (g : β ≃+* γ) :
...
   rfl
 
+open MulOpposite in
+/--
+For any ring `R`, we have ring isomorphism `Matₙₓₙ(Rᵒᵖ) ≅ (Matₙₓₙ(R))ᵒᵖ` given by transpose.
+-/
+@[simps apply symm_apply]
+def mopMatrix : Matrix m m αᵐᵒᵖ ≃+* (Matrix m m α)ᵐᵒᵖ where
+  toFun := fun M => op (M.transpose.map unop)
+  invFun := fun M => M.unop.transpose.map op
```
Comment (alreadydone) [inline Mathlib/Data/Matrix/Basic.lean:1507]:
```suggestion
  toFun M := op (M.transpose.map unop)
  invFun M := M.unop.transpose.map op
```

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean [hunk 2]
```diff
@@ -984,3 +984,85 @@ lemma end_apply_apply (ij : ι × ι) (k : ι) : (b.end ij) (b k) = if ij.2 = k
...
+section
+
+variable (ι : Type*) [Fintype ι] [DecidableEq ι]
+variable (R : Type*) [Semiring R]
+variable (M : Type*) [AddCommMonoid M] [Module R M]
+
+/--
+Let `M` be an `R`-module. Every `R`-linear map `Mⁿ → Mⁿ` corresponds to a `n×n`-matrix whose entries
+are `R`-linear maps `M → M`. In another word, we have`End(Mⁿ) ≅ Matₙₓₙ(End(M))` defined by:
+`(f : Mⁿ → Mⁿ) ↦ (x ↦ f (0, ...,x at j-th position, ..., 0) i)ᵢⱼ` and
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/Matrix/ToLin.lean:997]:
```suggestion
`(f : Mⁿ → Mⁿ) ↦ (x ↦ f (0, ..., x at j-th position, ..., 0) i)ᵢⱼ` and
```

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean [hunk 3]
```diff
@@ -984,3 +984,85 @@ lemma end_apply_apply (ij : ι × ι) (k : ι) : (b.end ij) (b k) = if ij.2 = k
...
+`m : Matₙₓₙ(End(M)) ↦ (v ↦ ∑ⱼ mᵢⱼ(vⱼ))`.
+
+See also `LinearMap.toMatrix'`
+-/
+@[simps]
+def endVecEquivMatrixEnd :
+    Module.End R (ι → M) ≃ Matrix ι ι (Module.End R M) where
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/Matrix/ToLin.lean:1004]: Why not inlining this into the RingEquiv?

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean [hunk 4]
```diff
@@ -984,3 +984,85 @@ lemma end_apply_apply (ij : ι × ι) (k : ι) : (b.end ij) (b k) = if ij.2 = k
...
+    ext i x j
+    simp only [LinearMap.coe_mk, AddHom.coe_mk, coe_comp, coe_single, Function.comp_apply]
+    rw [← Fintype.sum_apply, ← map_sum]
+    exact congr_arg₂ _ (by aesop) rfl
+  right_inv m := by ext; simp [Pi.single_apply, apply_ite]
+
+/--
+Let `M` be an `R`-module. Every `R`-linear map `Mⁿ → Mⁿ` corresponds to a `n×n`-matrix whose entries
+are `R`-linear maps `M → M`. In another word, we have`End(Mⁿ) ≅ Matₙₓₙ(End(M))` defined by:
+`(f : Mⁿ → Mⁿ) ↦ (x ↦ f (0, ...,x at j-th position, ..., 0) i)ᵢⱼ` and
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/Matrix/ToLin.lean:1023]:
```suggestion
`(f : Mⁿ → Mⁿ) ↦ (x ↦ f (0, ..., x at j-th position, ..., 0) i)ᵢⱼ` and
```

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean [hunk 5]
```diff
@@ -984,3 +984,85 @@ lemma end_apply_apply (ij : ι × ι) (k : ι) : (b.end ij) (b k) = if ij.2 = k
...
+`m : Matₙₓₙ(End(M)) ↦ (v ↦ ∑ⱼ mᵢⱼ(vⱼ))`.
+
+See also `LinearMap.toMatrix'`
+-/
+@[simps! apply_apply symm_apply_apply]
+def endVecAlgEquivMatrixEnd :
+    Module.End R (ι → M) ≃ₐ[R] Matrix ι ι (Module.End R M) where
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/Matrix/ToLin.lean:1058]: Can you do this with `[Algebra R A] [IsScalarTower R A M]`?
```suggestion
    Module.End A (ι → M) ≃ₐ[R] Matrix ι ι (Module.End A M) where
```

**Review by alreadydone — APPROVED** [review]

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean (hunk 5 above)
Comment (jjaassoonn) [reply Mathlib/LinearAlgebra/Matrix/ToLin.lean:1058]: Yes! Thanks

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean (hunk 3 above)
Comment (jjaassoonn) [reply Mathlib/LinearAlgebra/Matrix/ToLin.lean]: Thanks

**Comment by alreadydone** [comment]
Thanks!
maintainer merge

**Review by riccardobrasca — COMMENTED** [review]
Thanks!

bors merge

**Pull request closed without merging**